Bayesian - More verbose breaking text with worked examples (#24470)
home-assistant/home-assistant.github.io: source/_posts/2022-10-05-release-202210.markdown

{% details "Bayesian" %}

> As of 2022.10 `prob_given_false` is now a required configuration variable. Previously it had a default value of `1 - prob_given_true`, this didn't really have any mathematical basis and so has been removed. You can use this documentation to calculate what it should be, or to restore your previous functionality set it to `1 - prob_given_true`.

Reviewer: ```suggestion
`prob_given_false` is now a required configuration variable. Previously it had a default value of `1 - prob_given_true`; this didn't really have any mathematical basis and so has been removed. You can use this documentation to calculate what it should be, or to restore your previous functionality, set it to `1 - prob_given_true`.
```

> In 2022.10 the logic of Bayesian sensors was significantly corrected. Previously the configs given in the examples above would not have worked as expected. For example if the device `device_tracker.paulus` was `home` this would not have updated the probability that `Paulus Home` was on! This is because Bayesian ignored observations that were false, even though that *should* update our probabilities. 

Reviewer: ```suggestion
The logic of Bayesian sensors has been significantly corrected. Previously the configs given in the examples above would not have worked as expected. For example, if the device `device_tracker.paulus` was `home`, this would not have updated the probability that `Paulus Home` was on. This is because Bayesian ignored observations that were false, even though that *should* update our probabilities. 
```

> Because of this, many users will have had to use either of two workarounds:
> 1. Providing additional observations that evaluate to `True` when the other evaluates to `False`, effectively mirroring it. This forced Bayesian to take into account a negative observation.
> 2. Tweaking `prior`, `threshold` and even sometimes `prob_given_true` and `prob_given_false` to get the desired functionality, often with the help of a community provided spreadsheet for iterative testing.

Reviewer: ```suggestion
Because of this, many users will have had to use either of two workarounds:

1. Providing additional observations that evaluate to `True` when the other evaluates to `False`, effectively mirroring it. This forced Bayesian to take into account a negative observation.
2. Tweaking `prior`, `threshold`, and even sometimes `prob_given_true` and `prob_given_false` to get the desired functionality, often with the help of a community-provided spreadsheet for iterative testing.
```

> To solve (1) all you need do is delete the mirrored entry as shown below:

Reviewer: ```suggestion
To solve (1), all you need do is delete the mirrored entry as shown below:

```

```yaml
# Example of a mirrored entry used a workaround
binary_sensor:

>   name: "Heat On"

Reviewer: This whole block should be a in a list to match our YAML documentation standards.

```suggestion
  - name: "Heat On"
```

>   platform: "bayesian"

Reviewer: ```suggestion
  platform: bayesian
```

  prior: 0.2
  probability_threshold: 0.9
  observations:

>     - platform: "numeric_state"

Reviewer: ```suggestion
    - platform: numeric_state
```

>       entity_id: "sensor.outside_air_temperature_fahrenheit"

Reviewer: ```suggestion
      entity_id: sensor.outside_air_temperature_fahrenheit
```

      prob_given_true: 0.95
      prob_given_false: 0.05
      below: 50

>     - platform: "numeric_state" # line no longer needed - delete
>       entity_id: "sensor.outside_air_temperature_fahrenheit" # line no longer needed - delete

Reviewer: ```suggestion
    - platform: numeric_state # line no longer needed - delete
      entity_id: sensor.outside_air_temperature_fahrenheit # line no longer needed - delete
```

>       entity_id: "sensor.outside_air_temperature_fahrenheit" # line no longer needed - delete

Reviewer: The comments behind it might be too long to be readable on, for example, mobile devices.

      prob_given_true: 0.05 # line no longer needed - delete
      prob_given_false: 0.95 # line no longer needed - delete
      above: 50 # no longer needed - delete

>     - platform: "state"
>       entity_id: "binary_sensor.house_occupied"

Reviewer: ```suggestion
    - platform: state
      entity_id: binary_sensor.house_occupied
```

      prob_given_true: 0.3
      prob_given_false: 0.05
      to_state: "on"
    - platform: "state" # line no longer needed - delete

>       entity_id: "binary_sensor.house_occupied" # line no longer needed - delete

Reviewer: ```suggestion
    - platform: state # line no longer needed - delete
      entity_id: binary_sensor.house_occupied # line no longer needed - delete
```

      prob_given_true: 0.7 # line no longer needed - delete
      prob_given_false: 0.95 # line no longer needed - delete
      to_state: "off" # line no longer needed - delete
```

> To solve (2) you need to re-estimate your probabilities using this documentation - this will likely result in better performance in the long term. 

Reviewer: ```suggestion
To solve (2), you need to re-estimate your probabilities using this documentation - this will likely result in better performance in the long term.
```

> A hacky and *not recommended* workaround to restore your previous functionality would be to put in a mirrored entry for each observation where `prob_given_false` is equal to `prob_given_true`. This will induce the same behavior as before - the negative, counterfactual observation is essentially ignored. Be warned - this may break or cause errors in future versions and will not work for `template` or `numeric_state` observations. For templates, instead make them return `None` where they would have returned `False` to induce Bayesian to ignore them.

Reviewer: ```suggestion
A hacky and *not recommended* workaround to restore your previous functionality would be to put in a mirrored entry for each observation where `prob_given_false` is equal to `prob_given_true`. This will induce the same behavior as before - the negative, counterfactual observation is essentially ignored. Be warned - this may break or cause errors in future versions and will not work for `template` or `numeric_state` observations. For templates, instead, make them return `None` where they would have returned `False` to induce Bayesian to ignore them.
```

```yaml
binary_sensor:

>   name: "Heat On"

Reviewer: Same as above, while this works, `binary_sensor` is a list, and we should show it as such. (The whole block needs to be indented because of this suggested change).

```suggestion
  - name: "Heat On"
```

>   platform: "bayesian"

Reviewer: ```suggestion
  platform: bayesian
```

  prior: 0.2
  probability_threshold: 0.9
  observations:
    - platform: "state"

>       entity_id: "binary_sensor.house_occupied"

Reviewer: ```suggestion
    - platform: state
      entity_id: binary_sensor.house_occupied
```

      prob_given_true: 0.3
      prob_given_false: 0.05
      to_state: "on"

>     - platform: "state" # not 
>       entity_id: "binary_sensor.house_occupied" # not recommended

Reviewer: ```suggestion
    - platform: state # not 
      entity_id: binary_sensor.house_occupied # not recommended
```

2. Tweaking `prior`, `threshold`, and even sometimes `prob_given_true` and `prob_given_false` to get the desired functionality, often with the help of a community-provided spreadsheet for iterative testing.

To solve (1), all you need do is delete the mirrored entry as shown below:

```yaml
# Example of a mirrored entry used a workaround
binary_sensor:
  - name: "Heat On"
    platform: bayesian
    prior: 0.2
    probability_threshold: 0.9
    observations:
      - platform: numeric_state
        entity_id: sensor.outside_air_temperature_fahrenheit
        prob_given_true: 0.95
        prob_given_false: 0.05
        below: 50
      - platform: numeric_state # line no longer needed - delete
        entity_id: sensor.outside_air_temperature_fahrenheit # line no longer needed - delete
        prob_given_true: 0.05 # line no longer needed - delete
        prob_given_false: 0.95 # line no longer needed - delete
        above: 50 # no longer needed - delete
      - platform: state
        entity_id: binary_sensor.house_occupied
        prob_given_true: 0.3
        prob_given_false: 0.05

>         to_state: "on"

Author: ```suggestion
        to_state: on
```
should these quotes go as well?

      - platform: state # line no longer needed - delete
        entity_id: binary_sensor.house_occupied # line no longer needed - delete
        prob_given_true: 0.7 # line no longer needed - delete
        prob_given_false: 0.95 # line no longer needed - delete

>         to_state: "off" # line no longer needed - delete

Author: ```suggestion
        to_state: off # line no longer needed - delete
```
should these quotes go as well?

```

To solve (2), you need to re-estimate your probabilities using this documentation - this will likely result in better performance in the long term.

A hacky and *not recommended* workaround to restore your previous functionality would be to put in a mirrored entry for each observation where `prob_given_false` is equal to `prob_given_true`. This will induce the same behavior as before - the negative, counterfactual observation is essentially ignored. Be warned - this may break or cause errors in future versions and will not work for `template` or `numeric_state` observations. For templates, instead, make them return `None` where they would have returned `False` to induce Bayesian to ignore them.

```yaml
binary_sensor:

>   - name: "Heat On"

Author: ```suggestion
  - name: Heat On
```
?

    platform: bayesian
    prior: 0.2
    probability_threshold: 0.9
    observations:
      - platform: state
        entity_id: binary_sensor.house_occupied
        prob_given_true: 0.3
        prob_given_false: 0.05

>         to_state: "on"

Author: ```suggestion
        to_state: "on"
```
should these quotes go as well?

Reviewer: See answer above.

      - platform: state # not 
        entity_id: binary_sensor.house_occupied # not recommended
        prob_given_true: 0.5 # not recommended
        prob_given_false: 0.5 # not recommended

>         to_state: "off" # not recommended

Author: ```suggestion
        to_state: off # not recommended
```
should these quotes go as well?

Reviewer: No, that will cause problems in YAML (string vs boolean).
